File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
       

        sub = [] 
        for n in nums:
            
            i = bisect_left(sub, n)
            
            if i==len(sub):
                sub.append(n)
            
            else:
                sub[i] = n 
        
        return len(sub)

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        
         
    vector<int> sub ; 
    for(auto n:nums){
        auto ix = lower_bound(sub.begin() , sub.end() , n) ;

        if (ix == sub.end())
            sub.push_back(n) ; 

        else *ix = n  ;  

    }
    return  sub.size() ; 
    
        
        
    }
};